extensions/khr/GLSL_KHR_cooperative_matrix.txt

```diff
@@ -0,0 +1,396 @@
+
+The behavior of gl_MatrixOperandsSaturatingAccumulation and whether it is supported
+is documented in the Vulkan specification.
+
+Modify Section 9, Shading Language Grammar for Core Profile
+(Add to tokens list)
+COOPMAT
+
+(modify type_specifier to add type_parameter_specifier_opt)
+
+type_specifier:
+type_specifier_nonarray type_parameter_specifier_opt
+type_specifier_nonarray type_parameter_specifier_opt array_specifier
+
+(new rules)
+
+type_parameter_specifier_opt:
+type_parameter_specifier
+/*empty*/
+
+type_parameter_specifier:
+LEFT_ANGLE type_parameter_specifier_list RIGHT_ANGLE
+
+type_parameter_specifier_element:
+type_specifier
+unary_expression
+
+type_parameter_specifier_list:
+type_parameter_specifier_element
+type_parameter_specifier_list COMMA type_parameter_specifier_element
+
+
+Interactions with GL_EXT_shader_explicit_arithmetic_types_float16
+
+If GL_EXT_shader_explicit_arithmetic_types_float16 is not supported,
+remove the coopMatLoad/coopMatStore overloads that use float16_t.
+
+Issues
+
+(1) What are the grammar rules for type parameters?
+
+DISCUSSION: C++ template syntax has a parsing problem, because the
+rules allow a "conditional_expression" for the template parameters,
+which creates an ambiguity (shift/reduce conflict) where the parser
+can't easily tell whether a '>' is a greater-than operator or the end
+of the type parameter list. This means it's hard to parse something
+like
+
+coopmat<float16, gl_ScopeSubgroup, 16, A>B?16:8, gl_MatrixUseAccumulator>
+
+because it's unclear that the columns parameter is a ternary expression
+without looking ahead. The obvious way to make this example more clear
+is to add parentheses:
+
+coopmat<float16, gl_ScopeSubgroup, 16, (A>B?16:8), gl_MatrixUseAccumulator>
+
+This can be parsed as a "unary_expression" rather than
+"conditional_expression", and doesn't really lose any flexibility
+because unary_expression indirectly includes the pretty general
+
+"LEFT_PAREN expression RIGHT_PAREN" rule.
+
+RESOLVED: We diverge from the C++ grammar and use unary_expression
+for type parameters rather than conditional_expression.
+
+(2) What alignment rules should we have for buf/element/stride parameters
+in the load/store built-in functions?
+
+RESOLVED: The Vulkan SPIR-V environment appendix is responsible for
+documenting this. To summarize, the start of the matrix and the stride
+must be at least as aligned as the smaller of 16B or the size of a
+row/column of the matrix.
+
+(3) For the load/store functions, can the component type mismatch the array
+element type?
+
+RESOLVED: Yes, this makes it easier to efficiently load matrix data into
+shared memory. The stride parameter is interpreted in units of the
+pointed-to type, not in units of the matrix's component type. This
+extension includes overloads for 8 through 64-bit integers, and
+uvec2/uvec4.
+
+
+Revision History
+
+Revision 1
+
+- Internal revisions.
+
+Revision 2
```

Comment:
I suggest to delete rev2-3 and only document revisions that occur after publication.